File: api/app/domain/external/task.py
from typing import Optional
from app.domain.external.message_queue import MessageQueue
from typing import Protocol
from abc import ABC, abstractmethod


class TaskRunner(ABC):
    @abstractmethod
    async def run(self, task: Task):
        raise NotImplementedError

    @abstractmethod
    async def destroy(self):
        raise NotImplementedError

    @abstractmethod
    async def on_done(self, task: Task):
        raise NotImplementedError


class Task(Protocol):
    async def run(self) -> None: ...

    def cancel(self) -> None: ...

    @property
    def input_stream(self) -> MessageQueue: ...

    @property
    def output_stream(self) -> MessageQueue: ...

    @property
    def id(self) -> str: ...

    @property
    def done(self) -> bool: ...

    @classmethod
    def get(cls, task_id: str) -> Optional[Task]: ...

    @classmethod
    def create(cls, taskRunner: TaskRunner) -> Task: ...

    @classmethod
    def destroy(cls, task_id: str) -> None: ...

File: api-node/src/domain/external/task.ts
/** TaskRunner 与 Task 契约（对齐 Python app/domain/external/task.py） */

export interface TaskMessageQueue {
  put(message: string): Promise<string>;
  get(startId: string, blockMs: number): Promise<[string | null, string | null]>;
  pop(): Promise<[string | null, string | null]>;
  isEmpty(): Promise<boolean>;
}

export interface AgentTaskHandle {
  readonly id: string;
  readonly done: boolean;
  readonly inputStream: TaskMessageQueue;
  readonly outputStream: TaskMessageQueue;
  isCancelled(): boolean;
  run(): Promise<void>;
  cancel(): void;
}

export interface TaskRunner {
  readonly taskId: string;
  invoke(task: AgentTaskHandle): Promise<void>;
  destroy?(): Promise<void>;
  onDone?(task: AgentTaskHandle): Promise<void>;
}

export interface AgentTaskClass {
  create(runner: TaskRunner): AgentTaskHandle;
  get(taskId: string): AgentTaskHandle | undefined;
  destroy(): Promise<void>;
}
